### ISSUE TITLE
Setup concourse pipeline to handle Airbyte connections

### ISSUE THREAD

**Issue by shaidar** [mitodl/ol-data-platform#271]
Need to setup concourse pipelines to use the Octavia CLI and setup airbyte sources, destinations and connections

**Comment by shaidar** [comment]
The pipeline requires the following:
- Download the octavia-cli docker image. The cli is mainly a python package, however Airbyte only distributes it as a docker image.
- Sync state with S3, thus a pipeline requirement is an S3 resource. After some testing, we realized that the octavia-cli maintains state in yaml files and if those files aren't around, it'll recreate the same resources on Airbyte. So, the idea is to store state in S3 and sync it every time the cli runs.
- Run the cli with proper commands to setup sources, destinations, and connections
- Sync state again

Some of issues/problems I ran into so far:
- octavia-cli docker image doesn't use `latest` tag in docker hub and a specific version needs to be specified for the concourse pipeline to download the image
- reminding myself of concourse pipelines and their structure
- airbyte instance isn't accessible from concourse in addition to the auth complication.

**Comment by shaidar** [comment]
To connect concourse workers to the airbyte instance I was gonna temporarily add a sec group to allow access however when the issue I ran into is that the downloaded octavia-cli container doesn't have the consul client installed and thus couldn't resolve `airbyte.service.consul`. The workaround to that is to configure the consul client on the concourse worker to allow the traffic from containers by adding a `recursor` flag. However, after adding that and restarting the consul client, it was failing to restart. Initially I assumed it was an issue with the flag, so I removed it and tried starting the client again but it failed again. Looking through the logs, it appeared that it needs a bind address to be specified.

**Comment by shaidar** [comment]
Ran into several issues that took a bit to resolve:
- Had `-name` in output (no spacing) and while it wasn't flagged as an error and it kept updating the pipeline just fine, it wasn't using that key/value.
- Either not fully understanding how the [concourse-s3-sync-resource](https://github.com/mitodl/concourse-s3-sync-resource) works or there being an issue with the doc.
- Assumed the sync was failing because it wasn't downloading the folder from S3 to the docker image. The issue here turned out to be that since S3 is a key/value store, folders aren't really objects and so it had nothing to sync. Once I added a file in one of the folders, it was syncing it.
- Lack of online documentation to help with unblocking me.

Once I figured out the spacing and the s3 syncing issue things made a bit more sense and I was able to get the pipeline to spin up the octavia-cli docker image and sync the existing state from S3.

**Comment by shaidar** [comment]
Latest on the octavia-cli saga and the new roadblocks:
- For some reason and out of the blue, the pipeline hangs at trying to connect to the airbyte instance. It doesn't time out but just sits there trying to connect.
- Worked on the yaml configuration files that contain the connection details for the individual sources and destinations including the hostnames, usernames, and passwords of the DB's. As part of that, realized that we don't seem to have those hostnames resolvable in consul.
- Wasn't sure how we're going to be handling refreshing expired creds when setting up those connections.
  - Do we schedule the pipeline to run to refresh credentials? If so, then how will that work with our existing design to preserving state? The cli asks when it detects the presence of a `configuration.yaml` file whether to replace the existing resource, so maybe we should set it to `Y` and run the pipeline on a schedule.
  - Do we have access to the dynamic creds in concourse and if not what needs to happen to make that available? 
- Exclude the configuration files from being synced back to S3 as those would be rendered files and contains credentials. That can be done with additional options in the concourse-s3 resource, but need to have a naming convention for those files to exclude them and still sync state back up.
- Searching for docs on what the available options are under `replication_method` to set a schedule, pass in table names, and stream values.

**Comment by shaidar** [comment]
Didn't come across any docs on how/what goes into the `replication_method`, but imported existing sources/destination/connection configs that have been setup in the QA UI to take a look, and this is a sample `configuration.yaml` file:
```
definition_type: connection
resource_name: Django Postgres - MITx Online QA <> QA Data Lake
source_configuration_path: sources/django_postgres_mitx_online_qa/configuration.yaml
destination_configuration_path: destinations/qa_data_lake/configuration.yaml
configuration:
  sync_catalog:
    streams:
    - stream:
        name: courses_courserun
        json_schema:
          type: object
          properties:
            id:
              type: number
            live:
              type: boolean
            title:
              type: string
            run_tag:
              type: string
            end_date:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            course_id:
              type: number
            created_on:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            start_date:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            updated_on:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            courseware_id:
              type: string
            enrollment_end:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            expiration_date:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            enrollment_start:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            courseware_url_path:
              type: string
        supported_sync_modes:
        - full_refresh
        - incremental
        default_cursor_field: []
        source_defined_primary_key:
        - - id
        namespace: public
      config:
        sync_mode: incremental
        destination_sync_mode: append
        cursor_field:
        - id
        primary_key:
        - - id
        alias_name: courses_courserun
        selected: true
    - stream:
        name: courses_courserunenrollment
        json_schema:
          type: object
          properties:
            id:
              type: number
            active:
              type: boolean
            run_id:
              type: number
            user_id:
              type: number
            created_on:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            updated_on:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            edx_enrolled:
              type: boolean
            change_status:
              type: string
            enrollment_mode:
              type: string
            edx_emails_subscription:
              type: boolean
        supported_sync_modes:
        - full_refresh
        - incremental
        default_cursor_field: []
        source_defined_primary_key:
        - - id
        namespace: public
      config:
        sync_mode: incremental
        destination_sync_mode: append
        cursor_field:
        - id
        primary_key:
        - - id
        alias_name: courses_courserunenrollment
        selected: true
    - stream:
        name: users_user
        json_schema:
          type: object
          properties:
            id:
              type: number
            name:
              type: string
            email:
              type: string
            is_staff:
              type: boolean
            password:
              type: string
            username:
              type: string
            is_active:
              type: boolean
            created_on:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            last_login:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            updated_on:
              type: string
              format: date-time
              airbyte_type: timestamp_with_timezone
            is_superuser:
              type: boolean
        supported_sync_modes:
        - full_refresh
        - incremental
        default_cursor_field: []
        source_defined_primary_key:
        - - id
        namespace: public
      config:
        sync_mode: incremental
        destination_sync_mode: append
        cursor_field:
        - id
        primary_key:
        - - id
        alias_name: users_user
        selected: true
  status: active
  namespace_definition: customformat
  namespace_format: mitxonline/${SOURCE_NAMESPACE}
  prefix: mitxonline__app__postgres__
  schedule:
    units: 24
    time_unit: hours
```

**Comment by shaidar** [comment]
We decided to table this issue for now and use the Airbyte UI to handle things due to the following issues that would need to be resolved as they serve as blockers here:
- Lack of service discovery of the DB's to sync
- Access to vault credentials to connect to the DB's
- Figuring out how to best generate the json_schema in the `configuration.yaml` file

**Comment by blarghmatey** [comment]
Dagster now has support for managing Airbyte sources, destinations, and connectors as code. Closing this in favor of that approach.
